### PR TITLE
feat(orval): add output.artifacts for runtime-safe artifact output groups

### LINKED ISSUE #3704: feat(core): artifact output groups with runtime-safe, per-group entry points
## What's missing

One OpenAPI document can produce artifacts for three different runtimes: schemas and types (universal), Angular `HttpClient` / `httpResource` clients (browser), and MSW handlers plus Faker factories (Node test runners, Storybook, browser mocks). Orval can already write these to separate directories, but it has no first-class notion of an **artifact group**: a named bundle of generated files with its own barrel, safe to import without dragging in the other groups' runtime code. Two concrete holes remain, and one config still cannot describe more than one group boundary.

This is not Angular-specific. Angular is just where it hurts most, because a generated service imports `@angular/core` and evaluates an `@Injectable` decorator at module load (`samples/angular-app/src/api/http-client/pets/pets.service.ts:7-15`).

## Why it matters

In a monorepo, an app package re-exports generated output through one barrel:

```ts
export * from './lib/__generated__/openapi';
```

A Node-only mock package then imports a single handler:

```ts
import { getGetArticleMockHandler } from '@acme/reference-models';
```

Node evaluates the whole barrel, which evaluates the Angular client modules too. Under Playwright or Node instrumentation this can fail while Angular initializes platform code — even though the consumer only wanted an MSW handler. The reverse also bites: an Angular app that imports the client should not pull `@faker-js/faker` and `msw` into its module graph.

Note this is a **value**-level problem, not a type-only-export problem. Type imports are already safe: under `verbatimModuleSyntax`, `import type { A } from 'a'` is erased entirely from the emitted JS. What leaks is real runtime code.

## Today's behavior (verified on master @ 70fc4fad4)

Already works, and covers a good part of the original request:

- `schemas` accepts a path plus `importPath`, so types/Zod schemas already live in their own directory and can be imported through a package specifier — `packages/core/src/types.ts:330-334`.
- `mock` accepts a shared `path`, and each generator (`msw`, `faker`) accepts its own `path` — `packages/core/src/types.ts:530,554,578`.
- `mock.indexMockFiles: true` emits `index.msw.ts` / `index.faker.ts` mock-only barrels in `split` and `tags-split` — `packages/core/src/types.ts:570-577`, `packages/core/src/writers/split-tags-mode.ts:461-479`.
- Faker schema factories can be exposed through a dedicated barrel via `mock.generators[].schemasImportPath` — `packages/core/src/types.ts:539-544`.
- The workspace barrel already filters out every `*.<mockExt>.ts` file, so orval's own root barrel never re-exports MSW/Faker — `packages/orval/src/write-specs.ts:806-812` (shipped in `b059ca64e`, 2026-05-20).

Confirmed in `samples/angular-app`, which runs `mode: 'tags-split'` with `mock: { indexMockFiles: true, generators: [{ type: 'msw' }] }`. Result under `src/api/http-client/`: `pets/pets.service.ts`, `pets/pets.msw.ts`, `model/index.ts`, `index.msw.ts`.

Still broken:

1. **Mocks are inlined into the client file in `single` and `tags` modes.** De-inlining only happens when a mock `path` is set: `packages/core/src/writers/single-mode.ts:76` — `const shouldDeinlineMocks = hasAnyMockPath(output.mock);`. So by default, importing the generated client in those modes also evaluates `msw` and `@faker-js/faker`. This is the reported failure mode, and it needs no new config to fix.
2. **No client barrel in `tags-split`.** The root `index.ts` is only written when `tagsSplitDeduplication` is on: `packages/core/src/writers/split-tags-mode.ts:491` — `if (output.indexFiles && deduplicationEnabled && tagEntries.length > 0)`. Mocks get `index.msw.ts` but the clients get nothing. See `samples/angular-app/src/api/http-client/` — it has `index.msw.ts` and no `index.ts`.
3. **No group concept.** Splitting further (e.g. clients in one package, schemas in another, mocks in a third, each with its own barrel) means running several orval configs against the same input. That duplicates input config, operation naming, overrides, and post-processing, and lets groups drift apart.

## Proposed change

Split into two pieces. Ship the first on its own — it fixes the reported crash.

### Part A — de-inline mocks by default (small, no new config)

In `single` and `tags` modes, write mock code to sibling files (`<target>.msw.ts`, `<target>.faker.ts`) instead of appending it to the implementation file, matching what `split` and `tags-split` already do. Keep the current inlining reachable behind an explicit opt-in for anyone who depends on it.

Also emit the client root barrel in `tags-split` whenever `indexFiles` is true, not only when `tagsSplitDeduplication` is on.

### Part B — artifact groups with per-group entry points

The original proposal used `output.artifacts.<group>.{ target, entryPoint }`. That shape does not match orval's existing options and should be reworked:

- Sub-outputs in orval use `path`, not `target` (`schemas.path`, `mock.path`, `mock.generators[].path`). Only the top-level `output.target` is called `target`.
- There is no `entryPoint` concept. Barrels are `index*` files controlled by booleans (`output.indexFiles`, `output.mock.indexMockFiles`).
- `output.target`, `output.schemas`, and `output.mock.path` already carry three of the four group locations, so a parallel `artifacts` key would be a second way to say the same thing.

A shape that fits the existing options:

```ts
export default defineConfig({
  referenceApi: {
    input: { target: './openapi.yaml' },
    output: {
      mode: 'tags-split',
      target: './src/generated/angular',      // client group
      indexFiles: true,                        // -> ./src/generated/angular/index.ts
      schemas: {
        type: 'zod',
        path: './src/generated/schemas',       // schema group (already supported)
        importPath: '@acme/reference-models/zod',
      },
      mock: {
        indexMockFiles: true,                  // -> index.msw.ts / index.faker.ts
        generators: [
          { type: 'msw', path: './src/generated/msw' },
          { type: 'faker', path: './src/generated/faker' },
        ],
      },
    },
  },
});
```

Most of that already works today. What is new:

- `indexFiles: true` reliably produces a client-only `index.ts` in every split mode.
- That client barrel never re-exports mock or faker modules (already true of the workspace barrel; make it true of every group barrel).
- Cross-group imports inside generated files resolve through each group's `importPath` when one is set, so a mock file importing a schema goes to the schema package specifier, not a deep relative path.

Resulting layout:

```text
src/generated/
├── schemas/     index.ts        # Zod schemas + inferred types, no runtime deps
├── angular/     index.ts        # HttpClient / httpResource, imports @angular/*
├── msw/         index.msw.ts    # handlers, imports msw
└── faker/       index.faker.ts  # factories, imports @faker-js/faker
```

Consumers then map groups onto their own packages or ng-packagr secondary entry points (each secondary entry point is just a subdirectory with an `ng-package.json`) and get imports that state their runtime:

```ts
import type { Article } from '@acme/reference-models/zod';
import { getArticleResource } from '@acme/reference-data-access';
import { getGetArticleMockHandler } from '@acme/reference-mock';
```

### Out of scope

Orval should not generate Nx project config, `ng-package.json`, or `package.json` `exports` maps. Those stay with the host project. Orval's job is to emit the groups and their barrels so the host does not have to move files after every run.

## Acceptance

Part A:

- [ ] In `single` and `tags` modes with `mock` enabled and no mock `path` set, MSW and Faker code is written to separate files, not appended to the implementation file (`packages/core/src/writers/single-mode.ts`, `tags-mode.ts`).
- [ ] Importing the generated client file in Node does not evaluate `msw` or `@faker-js/faker`. Covered by a test.
- [ ] The previous inlined layout is still reachable through an explicit option, and the change is called out in the changelog as behavior-affecting.
- [ ] In `tags-split`, `indexFiles: true` writes the root client `index.ts` regardless of `tagsSplitDeduplication`.

Part B:

- [ ] Every group barrel (`index.ts`, `index.msw.ts`, `index.faker.ts`, schemas `index.ts`) re-exports only files from its own group.
- [ ] Generated cross-group imports honor each group's `importPath` when set.
- [ ] Operation names, type names, and overrides stay identical across groups from one config run.
- [ ] Existing single-target configs generate byte-identical output (default behavior unchanged apart from Part A).
- [ ] A sample under `samples/` builds all four groups from one config, with a Node-side test importing only the MSW barrel and asserting `@angular/core` is never loaded.
- [ ] `docs/content/docs/reference/configuration/output.mdx` documents the group model and links the existing `mock.path`, `mock.indexMockFiles`, `schemas.importPath`, and `mock.generators[].schemasImportPath` options as part of it.

## Related

- Angular DI-aware base URL composition (`override.angular.baseUrl`).
- Query-object serialization in Angular clients.
- MSW aliases missing from generated schema barrels.

### PR DESCRIPTION
**Pull request by the-ult**
> [!NOTE]
> **This PR is being retired (2026-08-05).** Current `master` already provides the runtime-safe artifact separation this proposed — verified by import-graph inspection, see the comment below. The two real remaining gaps are split out as **#3813** (Angular resource files honour `schemas.importPath` / `splitByTags`) and **#3815** (tags-split barrel re-exports `*.resource.ts` — the one genuinely-new capability here), with **#3814** as related hygiene. Kept open until those merge.

---

## Summary

Closes #3704.

Adds an **opt-in** `output.artifacts` option that emits independently importable artifact groups — Zod/TS **schemas**, the **client** (services + Angular `*.resource.ts`), and **msw**/**faker** mocks — each with its own group barrel, so a Node-based MSW consumer never evaluates Angular client modules (and vice versa).

```ts
output: {
  mode: 'tags-split',
  target: 'src/generated/client/api.ts',
  client: 'angular',
  artifacts: {
    schemas: 'src/generated/schemas',
    msw: 'src/generated/msw',
    faker: 'src/generated/faker',
  },
}
```

## Design: sugar over existing primitives, one new writer step

A triage pass over the codebase showed most of this already exists: `output.schemas` gives an independent schemas barrel, and `mock.generators[].path` + `indexMockFiles: true` already produce dedicated msw/faker dirs with `index.msw.ts` / `index.faker.ts` barrels. The genuine gap was a **complete client-group barrel** (in `tags-split` mode nothing re-exports the per-tag services together with Angular `*.resource.ts` extra files — the reason consumers hand-roll the unsafe root barrel this issue describes).

So the implementation is deliberately thin:

- `output.artifacts` **normalizes onto the existing options** (`output.schemas`, `mock.generators[].path`, forced `indexMockFiles: true`) — no parallel pipeline.
- One new writer step, `writeClientGroupBarrel` in `write-specs.ts`, emits `<clientDir>/index.ts` from the **actual** written implementation + `builder.extraFiles` paths (never derived from tag names), composing append-only with the `tagsSplitDeduplication` barrel.
- v1 requires `mode: 'tags-split'` and errors loudly on conflicts (`workspace`, `indexFiles: false`, duplicate schemas config, conflicting explicit generator paths, overlapping group dirs).
- The docs first show the **existing-primitives recipe** (usable today on any orval version), then `output.artifacts` as the ergonomic wrapper. If maintainers prefer docs-only + the client-barrel fix, the sugar layer is cleanly separable — happy to split the PR.

One real bug was caught by the generated-output typecheck gate during verification: Angular's httpResource generator duplicates helper boilerplate (`OrvalHttpResourceOptions`, `ResourceState`, `toResourceState`) into every `.resource.ts`, which makes naive `export *` barrels fail with TS2308. The barrel now re-exports those names explicitly from one canonical file before the wildcards (registry-based; extend `DUPLICATED_BOILERPLATE_EXPORTS_BY_CLIENT` if future generators duplicate boilerplate).

## Runtime-safety guarantees (tested)

New `tests/api-generation.spec.ts` assertions walk transitive relative imports and prove:

- msw/faker group entries never reach a `.service.ts`/`.resource.ts` file nor any `@angular/*` import;
- the client barrel never reaches `.msw.ts`/`.faker.ts` nor imports `msw`/`@faker-js/faker`;
- the client barrel re-exports both services and resources.

## Backwards compatibility

100% opt-in: `artifacts` is `undefined` by default and normalization is byte-identical without it. Full regeneration confirmed **zero modified existing snapshots** — 101 new files only (new test configs `artifactGroups`, `artifactGroupsZod`, client-agnostic `artifactGroupsFetch`, plus a new `samples/angular-app` artifact-groups target that CI compiles).

## Verification

- Unit: core 2100 ✓, orval 208 ✓ (1 pre-existing macOS-only `/tmp` flake), angular 222 ✓, mock 297 ✓
- Snapshots: 5375/5375 ✓ (non-update re-run green), angular-app sample 93/93 ✓
- `vp lint --type-aware --type-check` ✓; all 16 generated clients typecheck ✓

## Docs

- `docs/content/docs/guides/angular.mdx`: new “Runtime-safe artifact groups” section (problem, recipe with existing options, `artifacts` config, generated tree, monorepo `package.json#exports` mapping, caveats)
- `docs/content/docs/reference/configuration/output.mdx`: new `artifacts` reference

## Related issues

Sibling cluster: #3702, #3705, #3706, #3700. Prior art this builds on: #3537 (separate mock/faker output paths), #3318 (MSW barrel breaking vitest/jsdom — the failure mode fixed here), #3553 / #3624 (root barrels in split modes), #3592 (tags-split schemas), #2912 (indexFiles import paths), #3333 (rule-based output paths).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `output.artifacts` for generating independently importable groups (schemas, client, MSW, Faker) with per-group barrels in `tags-split` mode, including Angular per-tag client re-exports.
* **Documentation**
  * Expanded Angular “runtime-safe artifact groups” guidance and documented `output.artifacts` constraints/behavior; added tips to Faker and MSW guides.
* **Bug Fixes**
  * Improved cleanup to also remove generated files in configured MSW/Faker artifact directories.
* **Tests**
  * Added coverage for `output.artifacts` option normalization/validation, barrel generation behavior, and runtime import-graph separation.

### CODE REVIEW

**Comment by the-ult** [comment]
Review feedback addressed:

- **Copilot — `write-specs.test.ts` 'sorted' test name**: fixed in 894a52f0f (renamed; sorting intentionally lives in `writeClientGroupBarrel` via `.toSorted(localeCompare)`, the helper preserves caller order — also added newline-separator + dedupe guards to the append path with tests).
- **CodeRabbit — function-form `mock.generators` bypassing the artifacts conflict guard**: fixed in f9fd859e1 — the guard is hoisted before type-matching and rejects any function-form entry (its type/path can't be inspected), with a regression test.
- **CodeRabbit — sample `pets.resource.ts` reads signals outside the `httpResource` callback**: pre-existing generator behavior for the multi-content-type resource shape, byte-identical in the existing master samples (`samples/angular-app/src/api/http-resource/pets/pets.service.ts:202`, `http-resource-zod:211`) — the new artifact-groups sample just reproduces it. Real finding, but orthogonal to this PR; deserves its own issue against the Angular httpResource generator.
- **CodeRabbit — `preserveRequiredNullables` not passed in sample `pets.service.ts`**: likewise pre-existing — the identical 2-arg `filterParams` call ships in master's `samples/angular-app/src/api/http-client/pets/pets.service.ts:195`. Same suggestion: separate issue for the generator default.

**Comment by the-ult** [comment]
Broadened test coverage in d1418e0f0, per review discussion: added one `artifactGroups*` config per remaining client package (axios, react-query, swr, vue-query, svelte-query — Angular and fetch were already covered), parametrized the runtime-safety import-graph tests over all of them (now 12 tests total, +178 new snapshot files, zero diffs to any pre-existing snapshot), and cross-linked the feature from `msw.mdx`/`faker.mdx` since those are the docs that motivated the original runtime-safety concern.

`solid-start` was evaluated and deliberately left out: wiring it through this config is the first time solid-start-flavored output has ever run through `tests`' generated-output typecheck gate (its own `solid-start.config.ts` isn't part of `generate-api`'s `run-p 'generate:*'`), and doing so surfaces two pre-existing solid-start generator defects unrelated to `output.artifacts`: a missing `@solidjs/router` devDependency (TS2307) and a genuine `TS2395` merged-declaration error in generated `Pets` types — both reproducible from `solid-start.config.ts`'s own existing entries in isolation, nothing to do with this PR. Happy to file that as a separate issue if useful.

File: packages/orval/src/write-specs.test.ts [hunk 1]
```diff
@@ -84,3 +94,192 @@ describe('runFormatter', () => {
...
+
+  beforeEach(async () => {
+    workspace = await createTempWorkspace();
+  });
+
+  afterEach(async () => {
+    await rm(workspace, { recursive: true, force: true });
+  });
+
+  it('creates a new barrel file with sorted export lines', async () => {
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/orval/src/write-specs.test.ts]: Accepted - fixed in 894a52f0f. Renamed to `creates a new barrel file preserving the caller-provided order` (sorting happens one layer up, in `writeClientGroupBarrel` via `.toSorted(localeCompare)`); also added tests for the newline-separator and named-re-export-dedup edge cases the review led me to notice.

File: samples/angular-app/src/api/artifact-groups/client/pets/pets.resource.ts [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+  version?: Signal<number>,
+  options?:
+    | OrvalHttpResourceOptions<Pets, unknown>
+    | OrvalHttpResourceOptions<string, string>,
+): HttpResourceRef<Pets | string | undefined> {
+  const request = {
+    url: `/v${version?.() ?? 1}/pets`,
+    params: filterParams(params?.() ?? {}, new Set<string>([])),
+  };
+  const normalizedRequest: HttpResourceRequest = request;
```
[the remark this answers is not in the document]
Comment (the-ult) [reply samples/angular-app/src/api/artifact-groups/client/pets/pets.resource.ts]: Rejected for this PR - this is pre-existing Angular `httpResource` generator behavior for the multi-content-type (accept-header) resource shape, byte-identical in master's own committed sample at `samples/angular-app/src/api/http-resource/pets/pets.service.ts:202` and `http-resource-zod:211`. `output.artifacts` (this PR) only emits barrels from already-generated files - it doesn't touch how `packages/angular/src/http-resource.ts` builds request objects, so it can't fix this here even if it's a real bug. If it is genuinely wrong (worth confirming whether accept-header switching changes the signal-timing requirements vs. the single-content-type resources, which don't exhibit this pattern), it belongs in its own issue against the httpResource generator rather than this PR.

File: samples/angular-app/src/api/artifact-groups/client/pets/pets.service.ts [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+  ): Observable<AngularHttpResponse<TData>>;
+  searchPets<TData = Pets>(
+    params: SearchPetsParams,
+    version: number = 1,
+    options?: HttpClientObserveOptions,
+  ): Observable<TData | HttpEvent<TData> | AngularHttpResponse<TData>> {
+    const filteredParams = filterParams(
+      { ...params, ...options?.params },
+      new Set<string>(['requirednullableString', 'requirednullableStringTwo']),
+    );
```
[the remark this answers is not in the document]
Comment (the-ult) [reply samples/angular-app/src/api/artifact-groups/client/pets/pets.service.ts:196]: Rejected for this PR - filed as #3712 with full root-cause research, since it's more than a one-line fix. Confirmed pre-existing on plain master (`samples/angular-app/src/api/http-client/pets/pets.service.ts:195-197`, unmodified by this PR) and identical in `packages/angular/src/http-resource.ts:491` - `output.artifacts` only emits barrels from already-generated files and can't touch how `http-client.ts`/`http-resource.ts` build request objects.

Also worth noting: the suggested `+ true` diff doesn't actually work for this path. `preserveRequiredNullables` is gated on `!!paramsSerializer` by design (since the feature's original commit) because Angular's `HttpClient`/`httpResource` `params` field type structurally excludes `null` (`Record<string, string | number | boolean | ReadonlyArray<...>>`, see `@angular/common/http`); `filterParams(..., true)` returns a type that includes `null` and isn't assignable to `params:` without a serializer to consume it first - so applying the diff as-is would fail to compile. #3712 proposes a type-safe alternative (empty-string fallback instead of literal `null`) plus the exact code location.

File: samples/angular-app/src/api/artifact-groups/client/pets/pets.resource.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (the-ult) [reply samples/angular-app/src/api/artifact-groups/client/pets/pets.resource.ts]: Filed as #3713 with full root-cause research and a concrete fix proposal - it's a real bug (confirmed pre-existing on master), and broader than this one call site: it affects every generated multi-content-type `httpResource` (any verb, any params shape - path params like `petId()` included), since `accept`-based branch selection hoists the whole `request`/`headers` construction (and therefore every signal read needed to build it) out of any reactive `httpResource(() => ...)` factory context. Root cause + a closure-based fix that keeps the branch selection eager while restoring reactive signal reads are in the issue.

**Comment by melloware** [comment]
@aqeelat @wadakatu would like your opinion on this feature?

**Comment by aqeelat** [comment]
What's the use case? How are these artifacts going to be consumed by the generated clients? Can it be easily done via a post processor?

**Comment by melloware** [comment]
@aqeelat  sent you one more i forgot to send you: https://github.com/orval-labs/orval/security/advisories/GHSA-cxq5-97v7-87j8

**Comment by the-ult** [comment]
Artifacts might not be the correct naming.

The idea was in an nx repo.
We have a `data-access` folder/library and sometimes a `models` as well, to prevent Circular dependencies.
We als have a separate mocks library.
So it would be nice to be able to config where to generate the output.
And in in different folders/libraries if possible.
Or something like `data-access` with multiple secondary entry points (/data-access/msw, /data-access/models, data-access/services)

(Going on vacation now, will dive in it further when I get back. With improved examples, output, etc)

**Comment by the-ult** [comment]
Sorry for the radio silence — back from vacation now.

**Use case, concretely:** in an Nx-style monorepo you want independently-importable libraries so consumers don't pull unrelated runtime code across a boundary — e.g. a `data-access` lib whose consumers only need types/services, and a separate `mocks` lib that only test/storybook targets import. Today `tags-split` mode's only *complete* barrel is the workspace one, which re-exports everything — client code and mock code together. The moment a Node-based MSW handler imports that barrel (because it's the only complete entry point), it drags in Angular `HttpClient`/`httpResource` modules, and vice versa. That's the exact failure in #3318 (MSW barrel breaking vitest/jsdom), and what #3553/#3624/#3700/#3702/#3705/#3706 are all circling from different angles. So this isn't a new capability — it's making the existing split-mode file layout safely consumable without everyone hand-rolling (and getting wrong) their own root barrel.

**How consumers use it:** nothing changes for the generated client's own imports. What's new is a second, orthogonal barrel per artifact group — `<clientDir>/index.ts` re-exports every service/resource, `<schemasDir>/index.ts` re-exports schemas (already existed), `<mswDir>/index.msw.ts` / `<fakerDir>/index.faker.ts` re-export mocks (also already existed via `mock.generators[].path` + `indexMockFiles`). A consumer imports the barrel for what they need, and the runtime-safety tests in this PR prove those import graphs never cross.

**Why not a post-processor:** the barrel content isn't derivable from static config — it has to reflect what orval actually wrote (tags-split's file layout, Angular's extra `*.resource.ts` files, wherever `mock.generators[].path` sends mocks). A post-processor would have to re-walk the output directory and re-derive all of that from the outside — including the TS2308 dedup this PR already handles for Angular's httpResource boilerplate. Doing it in `write-specs.ts`, which already has the list of files it just wrote, is strictly less work and less fragile than re-deriving it externally.

**On the naming/Nx idea from before vacation:** I still think `artifacts` is the wrong word — it doesn't say *why* you'd reach for it. Directionally, mapping your `data-access`/`mocks`/`models` example onto today's config:

```ts
output: {
  mode: 'tags-split',
  target: 'libs/data-access/src/lib/generated/client/api.ts',
  client: 'angular',
  schemas: 'libs/models/src/lib/generated/schemas',
  artifacts: {
    msw: 'libs/mocks/src/lib/generated/msw',
    faker: 'libs/mocks/src/lib/generated/faker',
  },
}
```

produces:

```
libs/
├── data-access/
│   └── src/lib/generated/
│       └── client/
│           ├── index.ts              # import { listPets } from '@myorg/data-access'
│           ├── common-types.ts
│           └── pets/
│               ├── pets.service.ts
│               └── pets.resource.ts
├── models/
│   └── src/lib/generated/
│       └── schemas/
│           ├── index.ts              # import { Pet } from '@myorg/models'
│           ├── pet.ts
│           └── pets.ts
└── mocks/
    └── src/lib/generated/
        ├── msw/
        │   ├── index.msw.ts          # import { petsHandlers } from '@myorg/mocks/msw'
        │   └── pets/
        │       └── pets.msw.ts
        └── faker/
            ├── index.faker.ts        # import { getPetMock } from '@myorg/mocks/faker'
            └── pets/
                └── pets.faker.ts
```

Each folder is a normal Nx `lib` today, so `data-access` never imports MSW/Faker (nothing in it references `msw`/`@faker-js/faker`) and `mocks` never imports `@angular/*` — the runtime-safety tests in this PR assert exactly that boundary, per group. What's still missing to make each folder a *real* Nx secondary entry point is the scaffolding — `ng-package.json` / `project.json` / `package.json#exports` per lib — which is a one-time project-setup step, not something regenerated on every `orval` run; the docs section already shows the `package.json#exports` wiring for that by hand.

**Comment by aqeelat** [comment]
We recently added a couple of enhancements around this. In our nx repo, we do put schemas in a separate library than clients. We do use split-tags. I need to check if fakers and msw have the same configs.

Check recent releases and let me know if we're still not fully satisfying your requirements. If not, I'll help you get this PR to the finish line.

**Comment by the-ult** [comment]
@aqeelat Thanks — I checked, and you were right: recent releases do cover most of it. I verified the import graph on current `master` with `schemas.path` + `mock.generators[].path` + `indexMockFiles: true`, angular `tags-split`, `retrievalClient: 'both'`:

```
client/pets/pets.service.ts   client/pets/pets.resource.ts
models/index.ts  models/pet.ts  models/index.faker.ts
msw/index.msw.ts   msw/pets/pets.msw.ts
faker/index.faker.ts  faker/pets/pets.faker.ts
```

- `msw/index.msw.ts` → `pets.msw` imports `msw`, `../../faker/…`, and type-only `../../models`. **It never reaches `client/`, and there is no `@angular/*` anywhere in that graph.**
- `faker/index.faker.ts` → `pets.faker` imports `@faker-js/faker` and type-only `../../models`.
- `models/index.ts` is `export * from './pet'` only — it does **not** re-export `index.faker`, so the production type barrel stays free of faker.

So the runtime-safety core of #3704 — a Node MSW consumer never evaluating Angular — already works today. That's the answer to your question, and it's why I'm retiring this PR: `output.artifacts` was mostly sugar over primitives that now exist.

Two genuine gaps remained, both now standalone PRs:

- **#3813** — Angular's `*.resource.ts` ignored `schemas.importPath` (and `splitByTags`), emitting a deep relative cross-package path that `@nx/enforce-module-boundaries` rejects. That alone made `retrievalClient: 'both'` unusable from a library.
- **#3815** — in `tags-split`, `<clientDir>/index.ts` is the only complete client entry point but never re-exported the per-tag `*.resource.ts`. This is the one genuinely-new capability this PR contained (`writeClientGroupBarrel`), reduced to a targeted fix; the TS2308 boilerplate collision is handled by deriving the duplicate set from generated content rather than a registry.

Plus **#3814**, adjacent hygiene rather than a #3704 requirement: `clean` never pruned configured mock paths, so a separate mocks library accumulated orphaned `*.msw.ts` / `*.faker.ts` for endpoints the spec no longer has.

**One open question before I'd propose anything further.** With `faker: { path: './faker', schemas: true }`, the schema-level factories are written to `<schemas-dir>/index.faker.ts` — the *schemas* directory — regardless of `faker.path`. It isn't a safety leak (`models/index.ts` doesn't re-export it), but it does mean a models library's source imports `@faker-js/faker` while its public barrel never exposes it, which I'd expect to trip dependency-check lint rules in an Nx setup. `schemasImportPath` changes the specifier consumers import by, not where the file is written, so it doesn't help here. Is the current placement intentional — factories sitting next to the schemas they mock — or would honouring `faker.path` be welcome? Happy to do it if you'd take it; not going to build it speculatively.

Keeping this PR open until those three land, then closing it.
